packages/animations/lib/src/open_container.dart

```diff
@@ -0,0 +1,636 @@
+    Key key,
+    this.child,
+  }) : super(key: key);
+
+  final Widget child;
+
+  @override
+  State<_Hideable> createState() => _HideableState();
+}
+
+class _HideableState extends State<_Hideable> {
+  /// When non-null the child is replaced by a [SizedBox] of the set size.
+  Size get placeholderSize => _placeholderSize;
+  Size _placeholderSize;
+  set placeholderSize(Size value) {
+    if (_placeholderSize == value) {
+      return;
+    }
+    setState(() {
+      _placeholderSize = value;
+    });
+  }
+
+  /// When true the child is not visible, but will maintain its size.
+  ///
+  /// The value of this property is ignored when [placeholderSize] is non-null
+  /// (i.e. [isInTree] returns false).
+  bool get isVisible => _visible;
+  bool _visible = true;
+  set isVisible(bool value) {
+    assert(value != null);
+    if (_visible == value) {
+      return;
+    }
+    setState(() {
+      _visible = value;
+    });
+  }
+
+  /// Whether the child is currently included in the tree.
+  ///
+  /// When it is included, it may be visible or not according to [isVisible].
+  bool get isInTree => _placeholderSize == null;
+
+  @override
+  Widget build(BuildContext context) {
+    if (_placeholderSize != null) {
+      return SizedBox.fromSize(size: _placeholderSize);
+    }
+    return Opacity(
+      opacity: _visible ? 1.0 : 0.0,
+      child: widget.child,
+    );
+  }
+}
+
+class _OpenContainerRoute extends ModalRoute<void> {
+  _OpenContainerRoute({
+    @required Color closedColor,
+    @required this.openColor,
+    @required double closedElevation,
+    @required this.openElevation,
+    @required ShapeBorder closedShape,
+    @required this.openBuilder,
+    @required this.closedBuilder,
+    @required this.hideableKey,
+    @required this.closedBuilderKey,
+  })  : assert(closedColor != null),
+        assert(openColor != null),
+        assert(closedElevation != null),
+        assert(openElevation != null),
+        assert(closedShape != null),
+        assert(openBuilder != null),
+        assert(closedBuilder != null),
+        assert(hideableKey != null),
+        assert(closedBuilderKey != null),
+        _elevationTween = Tween<double>(
+          begin: closedElevation,
+          end: openElevation,
+        ),
+        _shapeTween = ShapeBorderTween(
+          begin: closedShape,
+          end: const RoundedRectangleBorder(),
+        ),
+        _colorTween = _FlippableTweenSequence<Color>(<TweenSequenceItem<Color>>[
+          TweenSequenceItem<Color>(
+            tween: ColorTween(begin: closedColor, end: Colors.white),
+            weight: 4 / 12,
+          ),
+          TweenSequenceItem<Color>(
+            tween: ColorTween(begin: Colors.white, end: openColor),
+            weight: 8 / 12,
+          ),
+        ]);
+
+  final Color openColor;
+  final double openElevation;
+  final OpenContainerBuilder closedBuilder;
+  final OpenContainerBuilder openBuilder;
+  final GlobalKey<_HideableState> hideableKey;
```

Comment:
It might help future readers of this code to explain what these global keys are for; here and in the constructor's parameters.